pkg/controller.v1/mpi/mpijob_controller.go

```diff
@@ -0,0 +1,1402 @@
+		Scheme:   mgr.GetScheme(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+		Log:      log.Log,
+	}
+
+	cfg := mgr.GetConfig()
+	kubeClientSet := kubeclientset.NewForConfigOrDie(cfg)
+	volcanoClientSet := volcanoclient.NewForConfigOrDie(cfg)
+	sharedInformers := informers.NewSharedInformerFactory(kubeClientSet, 0)
+	priorityClassInformer := sharedInformers.Scheduling().V1beta1().PriorityClasses()
+
+	r.JobController = common.JobController{
+		Controller:                  r,
+		Expectations:                expectation.NewControllerExpectations(),
+		Config:                      common.JobControllerConfiguration{EnableGangScheduling: enableGangScheduling},
+		WorkQueue:                   &util.FakeWorkQueue{},
+		Recorder:                    r.recorder,
+		KubeClientSet:               kubeClientSet,
+		VolcanoClientSet:            volcanoClientSet,
+		PriorityClassLister:         priorityClassInformer.Lister(),
+		PriorityClassInformerSynced: priorityClassInformer.Informer().HasSynced,
+		PodControl:                  control.RealPodControl{KubeClient: kubeClientSet, Recorder: r.recorder},
+		ServiceControl:              control.RealServiceControl{KubeClient: kubeClientSet, Recorder: r.recorder},
+	}
+	return r
+}
+
+// MPIJobReconciler reconciles a MPIJob object
+type MPIJobReconciler struct {
+	common.JobController
+	client.Client
+	Scheme   *runtime.Scheme
+	recorder record.EventRecorder
+	Log      logr.Logger
+}
+
+//+kubebuilder:rbac:groups=kubeflow.org,resources=mpijobs,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kubeflow.org,resources=mpijobs/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kubeflow.org,resources=mpijobs/finalizers,verbs=update
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=serviceaccount,verbs=get;list;watch;create;delete
+//+kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=roles,verbs=get;list;watch;create;delete;update
+//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=get;list;watch;create;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *MPIJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+	logger := r.Log.WithValues(mpiv1.Singular, req.NamespacedName)
+
+	mpijob := &mpiv1.MPIJob{}
+	err := r.Get(ctx, req.NamespacedName, mpijob)
+	if err != nil {
+		logger.Info(err.Error(), "unable to fetch MPIJob", req.NamespacedName.String())
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if err = validation.ValidateV1MpiJobSpec(&mpijob.Spec); err != nil {
+		logger.Info(err.Error(), "MPIJob failed validation", req.NamespacedName.String())
```

Comment:
why not stop reconciling?

Reply:
Agree. We return put the object back to the queue. It's worthless to move it forward. 
Another thing is I think the original validation logic is kind of outdated. Personally I don't see too many specific fields to validate. With latest controller tools, I think most of these invalid cases would be blocked by CRD validation.